src/storage/src/source/reclock.rs

```diff
@@ -8,413 +8,795 @@
+        // Ensure we have enough bindings and re-sync state if we lost a race
+        while let Err(Upper(global_upper)) = self.mint(&batch_upper).await {
+            self.sync(&global_upper).await;
+        }
 
-        let timestamp_bindings_listener = read_handle
-            .listen(as_of)
-            .await
-            .expect("Initial listen at handle.since ts")
-            .into_stream()
-            .boxed();
-
-        Ok(Self {
-            name,
-            read_progress,
-            write_upper,
-            persisted_timestamp_bindings,
-            read_cursors,
-            write_handle,
-            timestamp_bindings_listener,
-            now,
-            update_interval,
-            last_timestamp: Timestamp::minimum(),
+        Ok(ReclockIter {
+            reclock: self,
+            messages: batch.iter_mut(),
         })
     }
 
-    // TODO: require NowFn to be monotonic so we can simplify this function
-    fn get_time(&mut self) -> Option<Timestamp> {
-        let update_interval: u64 = self
-            .update_interval
-            .clone()
-            .as_millis()
-            .try_into()
-            .expect("interval millis don't fit into u64");
-        let new_ts = (self.now)();
-
-        // Mod and then round up
-        let remainder = new_ts % update_interval;
-        let new_ts_rounded = if remainder == 0 {
-            new_ts
-        } else {
-            new_ts - remainder + update_interval
-        };
-
-        if new_ts_rounded > self.last_timestamp {
-            self.last_timestamp = new_ts_rounded;
-            Some(new_ts_rounded)
-        } else {
-            None
+    /// Reclocks a `SourceTime` frontier into a `DestTime` frontier.
+    ///
+    /// The conversion has the property that all messages that are beyond the provided `SourceTime`
+    /// frontier will be relocked at times that will be beyond the returned `DestTime` frontier.
+    /// This can be used to drive a `DestTime` capability forward when the caller knows that a
+    /// `SourceTime` frontier has advanced.
+    ///
+    /// The method returns an error if the `SourceTime` frontier is not beyond the since frontier.
+    /// The error will contain the offending `SourceTime`.
+    pub fn reclock_frontier(
+        &self,
+        source_frontier: &HashMap<PartitionId, MzOffset>,
+    ) -> Result<Antichain<Timestamp>, (PartitionId, MzOffset)> {
+        // The upper is the greatest frontier that we can ever return
+        let mut dest_frontier = self.upper.clone();
+
+        let mut partitions = HashSet::new();
+        partitions.extend(self.source_upper.keys());
+        partitions.extend(source_frontier.keys());
+        // To refine it we have to go through all the partitions we know about and:
+        for pid in partitions {
+            let offset = source_frontier.get(pid).copied().unwrap_or_default();
+            // Ensure that the offsets are beyond the source since frontier
```

Comment:
This is a tricky bit and your comment here means I didn't do a good job explaining it in the comments.

The piece that's missing from the explanation is that there is no such thing as "partition you didn't ask for". All partitions are to be considered always present in a source frontier. What I mean by that is that the frontiers `HashMap::new()` and `HashMap::from_iter([(PartitionId::Kafka(1), MzOffset::from(0)), (PartitionId::Kafka(2), MzOffset::from(0))])` are exactly the same thing. Ideally we wouldn't be using a `HashMap` to represent the source frontier and use an actual `Antichain`. But because the set of partitions is very large the `Antichain` would have to have billions of zero entries.

In other words, the source frontier `HashMap`s are to be thought of as a special `Antichain` that absence of `PartitionId` implies the minimum offset is the upper for that partition id.

Back to your original question:

> As I understand it, if there's a partition the operator knows about but you didn't ask for (including in source_frontier), this will return an error only if you've compacted.

If a partition is in the `HashMap` of the operator then it means that its since is some non-zero value, otherwise it would be absent. So if someone calls `reclock_frontier` with a `HashMap` that doesn't include that partition **they are actually asking to reclock offset zero**, which we cannot do because we've compacted past it. So it's correct to return an error in this case.


> We could also add something similar if there's a partition the operator doesn't know about but you ask for? 

This is also handled the same way. In this case the partition since the operator knows about will be assumed to be zero (since it's absent) and you're trying to reclock some later offset for which a binding doesn't exist. In this case the function correctly ignores this in the calculation because the original estimate `self.upper` is valid since when the time comes to mint that binding its timestamp will surely be beyond `self.upper`.